src/noit_message_decoder.c

```diff
@@ -756,7 +756,10 @@ mtev_boolean add_tags_to_tagset_builder(noit_metric_tagset_builder_t *builder,
     } else if(next && (tag.total_size == tagstr_len || *next == ',')) {
       noit_metric_tagset_builder_el_t *el =
         (noit_metric_tagset_builder_el_t *) calloc(1, sizeof(noit_metric_tagset_builder_el_t));
-      el->tag = tag;
+      // copy tag to el->tag. may need to use the method from snowht to copy the
```

Comment:
```suggestion
      // copy tag to el->tag. may need to use the method from snowth to copy the
```

but this comment doesn't help, we can't possibly know which method you're talking about.  It would help if you can tell me what that is here please

Reply:
This PR is still in a draft state (it just doesn't allow me to set it to a draft). I pushed these changes for Michael to test reproducing for me last week, but have since determined the issue and just need to trace it up the stack. I'll ping you when this is cleaned up and ready for review.

Reply:
That's fine but looking at both PRs, I would like to introduce APIs to do this properly and also consolidate how these things are allocated.  This code is very unsafe and anything that modifies it must make it safer first

Reply:
Sounds good. Let's see what changes are left after I find the actual cause of the bug. Some of these changes may have been unnecessary and go away.